File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/RouteBuilder.java
package cn.wzbos.android.rudolph.router;


import android.os.Bundle;
import android.os.Parcelable;
import android.text.TextUtils;
import android.util.Log;
import android.util.SparseArray;

import java.io.Serializable;
import java.io.UnsupportedEncodingException;
import java.lang.reflect.Type;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import cn.wzbos.android.rudolph.Consts;
import cn.wzbos.android.rudolph.IRouteBuilder;
import cn.wzbos.android.rudolph.OnRouteListener;
import cn.wzbos.android.rudolph.RouteInfo;
import cn.wzbos.android.rudolph.RouteType;
import cn.wzbos.android.rudolph.Rudolph;
import cn.wzbos.android.rudolph.utils.TypeUtils;

/**
 * Bundle IBuilder
 * Created by wuzongbo on 2017/9/13.
 */
@SuppressWarnings("unchecked")
public abstract class RouteBuilder<B extends RouteBuilder<B, R>, R extends Router<?>> implements IRouteBuilder<RouteBuilder<B, R>, R> {


    Bundle args;

    OnRouteListener callback;

    Class<?> target;

    String rawUrl;

    String routePath;

    RouteType routeType;

    String routeTag;

    Map<String, Type> queryParameters;


    @Override
    public B putExtra(Bundle map) {
        args.putAll(map);
        return (B) this;
    }

    @Override
    public B putExtra(String key, String value) {
        args.putString(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, boolean value) {
        args.putBoolean(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, short value) {
        args.putShort(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, int value) {
        args.putInt(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, long value) {
        args.putLong(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, double value) {
        args.putDouble(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, byte value) {
        args.putByte(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, char value) {
        args.putChar(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, float value) {
        args.putFloat(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, CharSequence value) {
        args.putCharSequence(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, Parcelable value) {
        args.putParcelable(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, Parcelable[] value) {
        args.putParcelableArray(key, value);
        return (B) this;
    }

    @Override
    public B putParcelableArrayListExtra(String key, ArrayList<? extends Parcelable> value) {
        args.putParcelableArrayList(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, SparseArray<? extends Parcelable> value) {
        args.putSparseParcelableArray(key, value);
        return (B) this;
    }

    @Override
    public B putIntegerArrayListExtra(String key, ArrayList<Integer> value) {
        args.putIntegerArrayList(key, value);
        return (B) this;
    }

    @Override
    public B putStringArrayListExtra(String key, ArrayList<String> value) {
        args.putStringArrayList(key, value);
        return (B) this;
    }

    @Override
    public B putCharSequenceArrayListExtra(String key, ArrayList<CharSequence> value) {
        args.putCharSequenceArrayList(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, Serializable value) {
        args.putSerializable(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String name, boolean[] value) {
        args.putBooleanArray(name, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, byte[] value) {
        args.putByteArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, short[] value) {
        args.putShortArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, char[] value) {
        args.putCharArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, int[] value) {
        args.putIntArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, long[] value) {
        args.putLongArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, float[] value) {
        args.putFloatArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, double[] value) {
        args.putDoubleArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, String[] value) {
        args.putStringArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, CharSequence[] value) {
        args.putCharSequenceArray(key, value);
        return (B) this;
    }

    @Override
    public B putExtra(String key, Bundle value) {
        args.putBundle(key, value);
        return (B) this;
    }

    @Override
    public Bundle getExtras() {
        return args;
    }

    @Override
    public B onListener(OnRouteListener callback) {
        this.callback = callback;
        return (B) this;
    }

    private RouteBuilder() {
        this.args = new Bundle();
    }

    RouteBuilder(Class<?> target) {
        this();
        this.target = target;
    }


    RouteBuilder(String rawUrl) {
        this();
        this.rawUrl = rawUrl;

        if (!TextUtils.isEmpty(rawUrl)) {
            RouteInfo routeInfo = Rudolph.getRouter(getPath());
            if (null == routeInfo) {
                if (Rudolph.getRouters() == null || Rudolph.getRouters().size() == 0) {
                    Log.e("Rudolph", "错误：没有匹配到相关路由:" + rawUrl + ",当前路由表为空,请确认是否已进行 Rudolph.init() 初始化操作！");
                } else {
                    Log.e("Rudolph", "错误：没有匹配到相关路由:" + rawUrl);
                }
                return;
            }

            this.target = routeInfo.getTarget();
            this.routePath = routeInfo.getPath();
            this.routeTag = routeInfo.getTag();
            this.routeType = routeInfo.getRouteType();
            this.queryParameters = routeInfo.getParams();

            //返回的所有的地址参数与查询参数值
            Map<String, String> resultMap = getUriAllParams();

            if (resultMap != null) {
                //页面接收参数
                for (Map.Entry<String, String> kv : resultMap.entrySet()) {
                    String key = kv.getKey();
                    String val = kv.getValue();
                    addParams(key, val);
                }
            }
        }
    }

    private void addParams(String key, String val) {
        if (queryParameters != null && !queryParameters.isEmpty() && queryParameters.containsKey(key)) {
            Type type = queryParameters.get(key);
            if (!android.text.TextUtils.isEmpty(val)) {
                TypeUtils.getObject(null, key, val, type, this);
            }
        } else {
            args.putString(key, val);
        }
    }

    String getPath() {
        int n = rawUrl.indexOf("://");
        int stx, etx;
        if (n > -1) {
            stx = n + 3;
            etx = rawUrl.indexOf("?", stx);
        } else {
            stx = 0;
            etx = rawUrl.indexOf("?");
        }

        String path;
        if (etx > -1) {
            path = rawUrl.substring(stx, etx);
        } else {
            path = rawUrl.substring(stx);
        }
        return path;
    }

    private List<String> getSegments() {
        List<String> segments = new ArrayList<>();
        String[] values = getPath().split("/");
        for (String val : values) {
            try {
                if (!TextUtils.isEmpty(val)) {
                    segments.add(URLDecoder.decode(val, "utf-8"));
                }
            } catch (UnsupportedEncodingException e) {
                Log.e("rudolph", "getSegments failed!", e);
            }
        }
        return segments;
    }

    private String getEncodedQuery() {
        int n = this.rawUrl.indexOf("?");
        if (n > -1) {
            return this.rawUrl.substring(n + 1);
        } else {
            return null;
        }
    }

    /**
     * 获取当前URL地址的所有参数
     *
     * @return 返回参数kv集合
     */
    Map<String, String> getUriAllParams() {
        HashMap<String, String> params = new LinkedHashMap<>();

        String[] routeSegments = routePath.substring(1).split("/");

        List<String> pathSegments = getSegments();

        //segments个数不匹配
        if (routeSegments.length != pathSegments.size())
            return null;
        //raw uri
        params.put(Consts.RAW_URI, this.rawUrl);
        //path params
        for (int i = 0; i < routeSegments.length; i++) {
            if (routeSegments[i].startsWith(":")) {
                //put path params
                params.put(routeSegments[i].substring(1), pathSegments.get(i));
                continue;
            }

            if (!routeSegments[i].equalsIgnoreCase(pathSegments.get(i)))
                return null;
        }

        //put query params
        String query = getEncodedQuery();
        if (!TextUtils.isEmpty(query)) {
            String[] queryParameters = query.split("&");
            for (String str : queryParameters) {

                String[] kv = str.split("=");

                if (kv.length == 2) {

                    try {
                        String name = URLDecoder.decode(kv[0], "utf-8");
                        String value = URLDecoder.decode(kv[1], "utf-8");
                        params.put(name, value);
                    } catch (Exception e) {
                        Log.e("rudolph", "getUriAllParams failed!", e);
                    }
                }
            }
        }

        return params;
    }

}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/RouteBuilder.kt
package cn.wzbos.android.rudolph.router

import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.util.SparseArray
import cn.wzbos.android.rudolph.*
import cn.wzbos.android.rudolph.Rudolph.getRouter
import cn.wzbos.android.rudolph.logger.RLog
import cn.wzbos.android.rudolph.utils.TypeUtils
import java.io.Serializable
import java.net.URLDecoder
import java.util.*

/**
 * Bundle IBuilder
 * Created by wuzongbo on 2017/9/13.
 */
@Suppress("UNCHECKED_CAST")
abstract class RouteBuilder<B : RouteBuilder<B, R>?, R : Router<*>?>(val rawUrl: String) :
    IRouteBuilder<RouteBuilder<B, R>?, R> {
    companion object {
        const val TAG = "RouteBuilder"
    }

    override var extras: Bundle = Bundle()
    var scheme: MutableList<String>? = null
    var host: MutableList<String>? = null
    var path: MutableList<String>? = null

    var callback: OnRouteListener? = null
        private set
    var target: Class<*>? = null
        private set
    var routeType: RouteType? = null
        private set
    var routeTag: String? = null
        private set
    var extraTypes: MutableMap<String, ExtraType>? = null
        private set
    var interceptors: MutableList<Class<out RouteInterceptor>>? = null
        private set

    private var encodedQuery: String? = null


    override fun putExtra(map: Bundle?): B {
        extras.putAll(map)
        return this as B
    }

    override fun putExtra(key: String?, value: String?): B {
        extras.putString(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Boolean): B {
        extras.putBoolean(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Short): B {
        extras.putShort(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Int): B {
        extras.putInt(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Long): B {
        extras.putLong(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Double): B {
        extras.putDouble(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Byte): B {
        extras.putByte(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Char): B {
        extras.putChar(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Float): B {
        extras.putFloat(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: CharSequence?): B {
        extras.putCharSequence(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Parcelable?): B {
        extras.putParcelable(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Array<Parcelable?>?): B {
        extras.putParcelableArray(key, value)
        return this as B
    }

    override fun putParcelableArrayListExtra(
        key: String?,
        value: ArrayList<out Parcelable?>?,
    ): B {
        extras.putParcelableArrayList(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: SparseArray<out Parcelable?>?): B {
        extras.putSparseParcelableArray(key, value)
        return this as B
    }

    override fun putIntegerArrayListExtra(key: String?, value: ArrayList<Int?>?): B {
        extras.putIntegerArrayList(key, value)
        return this as B
    }

    override fun putStringArrayListExtra(key: String?, value: ArrayList<String?>?): B {
        extras.putStringArrayList(key, value)
        return this as B
    }

    override fun putCharSequenceArrayListExtra(
        key: String?,
        value: ArrayList<CharSequence?>?,
    ): B {
        extras.putCharSequenceArrayList(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Serializable?): B {
        extras.putSerializable(key, value)
        return this as B
    }

    override fun putExtra(name: String?, value: BooleanArray?): B {
        extras.putBooleanArray(name, value)
        return this as B
    }

    override fun putExtra(key: String?, value: ByteArray?): B {
        extras.putByteArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: ShortArray?): B {
        extras.putShortArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: CharArray?): B {
        extras.putCharArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: IntArray?): B {
        extras.putIntArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: LongArray?): B {
        extras.putLongArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: FloatArray?): B {
        extras.putFloatArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: DoubleArray?): B {
        extras.putDoubleArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Array<String?>?): B {
        extras.putStringArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Array<CharSequence?>?): B {
        extras.putCharSequenceArray(key, value)
        return this as B
    }

    override fun putExtra(key: String?, value: Bundle?): B {
        extras.putBundle(key, value)
        return this as B
    }

    override fun onListener(callback: OnRouteListener?): B {
        this.callback = callback
        return this as B
    }

    private fun putUriAllParams() {
        //返回的所有的地址参数与查询参数值
        val resultMap = uriAllParams
        for (kv: Map.Entry<String, String?> in resultMap.entries) {
            addParams(kv.key, kv.value)
        }
    }

    private fun addParams(key: String, value: String?) {
        if (extraTypes?.containsKey(key) == true) {
            val type = extraTypes?.get(key)
            if (!value.isNullOrEmpty() && type != null) {
                TypeUtils.getObject(null, key, value, type, this)
            }
        } else {
            //未用 @Extra 注解声明的参数类型，
            extras.putString(key, value)
        }
    }

    init {
        RLog.i(TAG, "rawUrl=${rawUrl}")

        if (rawUrl.isNotBlank()) {
            val uri = Uri.parse(rawUrl)
            uri.path?.let {
                this.path = mutableListOf(it)
            }

            encodedQuery = rawUrl.indexOf("?").let {
                if (it > -1) {
                    rawUrl.substring(it + 1)
                } else {
                    null
                }
            }
            val info = getRouter(rawUrl)
            if (info != null) {
                setRouterInfo(info)
            } else {
                RLog.e(TAG, "没有路由信息，rawUrl=${rawUrl}")
            }
            putUriAllParams()
        }
    }

    private fun setRouterInfo(it: RouteInfo) {
        val targetName = it.target
        if (!targetName.isNullOrEmpty()) {
            try {
                target = Class.forName(targetName)
            } catch (e: ClassNotFoundException) {
                RLog.e(TAG, "错误：路由类加载失败！$target", e)
            }
        } else {
            target = it.targetClass
        }
        routeTag = it.tag
        routeType = it.type
        extraTypes = it.extras
        interceptors = it.interceptors
    }


    /**
     * 获取当前URL地址的所有参数
     *
     * @return 返回参数kv集合
     */
    val uriAllParams: MutableMap<String, String?>
        get() {
            val params: MutableMap<String, String?> = LinkedHashMap()
            //raw uri
            params[Consts.RAW_URI] = rawUrl
            //put query params
            val query = encodedQuery
            if (!query.isNullOrEmpty()) {
                val queryParameters = query.split("&".toRegex()).toTypedArray()
                for (str: String in queryParameters) {
                    val kv = str.split("=".toRegex()).toTypedArray()
                    if (kv.size == 2) {
                        try {
                            val name = URLDecoder.decode(kv[0], "utf-8")
                            val value = URLDecoder.decode(kv[1], "utf-8")
                            params[name] = value.ifEmpty { null }
                        } catch (e: Exception) {
                            RLog.e(TAG, "getUriAllParams failed!", e)
                        }
                    }
                }
            }
            return params
        }

}